File: [23.12]Design-Pattern/Java/src/behavioral/state/Laptop.java
package behavioral.state;

public class Laptop {
    private PowerState powerState;

    public Laptop() {
        this.powerState = OffState.getInstance();
    }

    public void setPowerState(PowerState powerState) {
        this.powerState = powerState;
    }

    public void powerButtonPush(){
        powerState.powerButtonPush(this);
    }

    public void typeButtonPush(){
        powerState.typeButtonPush();
    }

    void currentStatePrint(){
        System.out.println(powerState.toString());
    }

}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/behavioral/state/Laptop.kt
package behavioral.state

class Laptop {
    private var state: State = OffState

    fun setState(state: State) {
        this.state = state
    }

    fun powerButtonPush() {
        state.powerButtonPush(this)
    }

    fun typeButtonPush() {
        state.typeButtonPush()
    }

    fun currentState() {
        println(state.toString())
    }


}